### PR TITLE
Add index_add op

### LINKED ISSUE #3638: [inductor] index_add / scatter_add scatter-combine returns SILENTLY WRONG results (no crash)
# [inductor] index_add / scatter_add scatter-combine returns SILENTLY WRONG results (no crash)

## Summary

After each routed row is processed by its expert, the K expert outputs for a token
must be summed back into that token's output row — an `index_add`/`scatter_add` from
`[N = T*K, H]` per-row outputs into `[T, H]`. On Spyre all three natural spellings
**compile and run without error but return wrong values**. This is the dangerous
mode: a naive MoE combine produces garbage with no signal.

## Repro

`repros/gemma4_moe/issues/p5_index_add_scatter_reduce.py` (no model load, no real weights):

```python
import torch, torch_spyre
T, K, H = 64, 4, 2816
N = T * K

tok = torch.arange(T).repeat_interleave(K)   # [N], each token appears K times
seg = torch.randn(N, H, dtype=torch.float16) # per-row expert outputs
out = torch.zeros(T, H, dtype=torch.float16) # token accumulator

def combine(out, tok, seg):
    out = out.clone(); out.index_add_(0, tok, seg); return out

ref = combine(out.clone(), tok.clone(), seg.clone())            # CPU reference
got = torch.compile(combine, dynamic=False)(
        out.clone().to("spyre"), tok.clone().to("spyre"), seg.clone().to("spyre")).cpu()
# torch.allclose(got, ref) is False
```

Also tested: functional `out.index_add(...)` and `out.scatter_add_(0, idx_expand, seg)`
— same result.

## Observed (on-device, 2026-08-10)

```
index_add_   : WRONG (mismatch CPU)
index_add    : WRONG (mismatch CPU)
scatter_add_ : WRONG (mismatch CPU)
```

Characterization of the `index_add_` mismatch:

```
rows matching CPU reference : 0 / 64
mean |out|  reference=1.592   device=0.797   (≈ half)
got == first-occurrence-only? False
got == zeros?                 False
```

Output magnitude is roughly **half** the reference and **0/T** rows are correct — the
device accumulates only some of each token's K contributions. It is not a recognizable
"first/last-writer-wins" or "dropped" pattern; it is a genuinely wrong scatter-reduce.

## Impact

The MoE combine step (summing K expert outputs per token) silently corrupts the
model output on device. Because there is no crash, this can pass smoke tests and only
show up as bad generations / accuracy loss.

## Suggested fix / discussion

- A correct on-device scatter-reduce for `index_add`/`scatter_add` with **repeated
  destination indices** (the whole point of the MoE combine), or
- until then, the grouped/tiled per-expert `index_add_` scheme the MoE RFC specifies
  (contiguous per-expert segments, sink-row padding) — but that still needs a correct
  underlying scatter-add.

At minimum, if repeated-index scatter-add cannot be lowered correctly, it should
**abort** rather than return wrong values silently.

## Labels

bug, inductor, torch-ops (complex)

### PR DESCRIPTION
**Pull request by shruthip-venkatesh**
#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Enables `aten.index_add` / `index_add_` on the Spyre backend :

**Custom decomposition (`decompositions.py`)**

Registers `spyre_index_add` for `aten.index_add.default`. The decomposition lowers
`index_add` to three primitives that Spyre's indirect-access engine already supports:

```
index_select(self, dim, index)      # on-device indirect gather
+ source * alpha                    # add (with optional alpha scaling)
index_put(self, [index], updated)   # on-device indirect overwrite scatter
```

This is an exact read-modify-write when `index` contains no duplicate values.
Duplicate indices are a precondition violation — the overwrite scatter keeps only
the last writer, producing silently wrong values. This matches the same constraint
documented in PyTorch's own `index_add` semantics for non-accumulating backends.

#### Which issue(s) this PR is related to:

Fixes #3638

#### Tests

- `tests/inductor/test_indirect_access_scatter.py` — 18 new `test_index_add_*` tests covering:
  - End-to-end value correctness (`expect_close=True`) for the MoE regression shape `[64, 2816]`, partial/dense/full updates, alpha scaling, fp32, ragged N, and the functional vs in-place API forms

#### Special notes for your reviewer:

- **Unique indices only.** The decomposition is a gather → add → overwrite-scatter (RMW). Duplicate indices produce wrong values silently — this is a data precondition, not a compiler invariant, and matches the behaviour of the existing `index_put` path.
- **`dim != 0` is unsupported.** `index_add` on `dim=1` crashes during compilation (gather leg puts the index-dependent coordinate on a non-outermost device dim; `propagate_layouts` finds no valid layout). Pinned as `CRASHED` in `test_index_add_dim1_unsupported`.

### CODE REVIEW

**Review by avery-blanchard — APPROVED** [review]
looks good to me
